File: app/src/main/java/com/tapura/podmorecasts/details/PodcastDetailsViewModel.java
package com.tapura.podmorecasts.details;


import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.util.Pair;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.tapura.podmorecasts.MyApplication;
import com.tapura.podmorecasts.MyLog;
import com.tapura.podmorecasts.cache.PodcastCache;
import com.tapura.podmorecasts.database.FirebaseDb;
import com.tapura.podmorecasts.model.Podcast;


import java.util.Objects;

public class PodcastDetailsViewModel extends ViewModel implements DownloadAndParseFeedTask.CallBack {

    private MutableLiveData<Podcast> mCurrentPodcast;
    private Podcast cachedPodcast;

    public MutableLiveData<Podcast> getCurrentPodcast(String feed) {
        if (mCurrentPodcast == null) {
            mCurrentPodcast = new MutableLiveData<>();

            cachedPodcast = new PodcastCache().getCache(MyApplication.getApp(), feed);
            if (cachedPodcast == null) {
                new DownloadAndParseFeedTask(this).execute(feed);
            }
            mCurrentPodcast.setValue(cachedPodcast);
        }
        return mCurrentPodcast;
    }

    @Override
    public void onDownloadAndParseFinished(Podcast podcast) {
        cachedPodcast = podcast;
        new PodcastCache().setCache(MyApplication.getApp(), podcast);
        mCurrentPodcast.setValue(cachedPodcast);
    }
}

File: app/src/main/java/com/tapura/podmorecasts/details/PodcastDetailsViewModel.kt
package com.tapura.podmorecasts.details


import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.util.Log
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.tapura.podmorecasts.MyApplication
import com.tapura.podmorecasts.cache.PodcastCache
import com.tapura.podmorecasts.database.FirebaseDb
import com.tapura.podmorecasts.model.Podcast

class PodcastDetailsViewModel : ViewModel() {

    private val database: FirebaseDb by lazy { FirebaseDb() }
    private val cache: PodcastCache by lazy { PodcastCache() }

    private lateinit var mCurrentPodcast: MutableLiveData<Podcast>

    fun getCurrentPodcast(feed: String): MutableLiveData<Podcast> {
        if (::mCurrentPodcast.isInitialized.not()) {
            mCurrentPodcast = MutableLiveData()

            val cachedPodcast = cache.getCache(MyApplication.getApp(), feed)

            if (cachedPodcast == null) {
                Log.d("thales", "cache null")
                DownloadAndParseFeedTask { podcast ->
                    podcast?.let { discoverIfItIsFav(it) }
                }.execute(feed)
            }

            mCurrentPodcast.postValue(cachedPodcast)
        }
        return mCurrentPodcast
    }

    private fun discoverIfItIsFav(podcast: Podcast) {
        FirebaseDb().getPodcast(MyApplication.getApp(),
                podcast.feedUrl,
                object : ValueEventListener {
                    override fun onCancelled(error: DatabaseError) {
                        Log.d("thales", "getPodcast cancelled, nothing to do, ret=$error")
                    }

                    override fun onDataChange(data: DataSnapshot) {
                        podcast.isFavorite = data.exists()
                        cache.setCache(MyApplication.getApp(), podcast)
                        mCurrentPodcast.postValue(podcast)
                    }
                })
    }

    fun setFav(isFav: Boolean) {
        mCurrentPodcast.value?.let {
            it.isFavorite = isFav
            mCurrentPodcast.postValue(it)

            cache.setCache(MyApplication.getApp(), it)
        }
    }
}
